dashboard/modules/runtime_env/runtime_env_agent.py

```diff
@@ -59,6 +59,104 @@ class CreatedEnvResult:
     result: str
 
 
+class UriType(Enum):
+    WORKING_DIR = 1
+    PY_MODULES = 2
+    PIP = 3
+    CONDA = 4
+
+
+class ReferenceTable:
+    """
+    The URI reference table which is used for GC.
+    When the reference count is decreased to zero,
+    the URI should be removed from this table and
+    added to cache if needed.
+    """
+
+    def __init__(
+        self,
+        uris_parser: Callable[[RuntimeEnv], None],
+        unused_uris_callback: Callable[[List[Tuple[UriType, str]]], None],
+        unused_runtime_env_callback: Callable[[str], None],
+    ):
+        self._logger = default_logger
+        self._uri_reference: Dict[str, int] = dict()
+        self._runtime_env_reference: Dict[str, int] = dict()
+        self._uris_parser = uris_parser
+        self._unused_uris_callback = unused_uris_callback
+        self._unused_runtime_env_callback = unused_runtime_env_callback
+        # Don't change URI reference for `client_server` because `client_server` doesn't
+        # send the `DeleteRuntimeEnvIfPossible` RPC when the client exits.
+        self._reference_exclude_sources: Set[str] = {
+            "client_server",
+        }
+
+    def _increase_reference_for_uris(self, uris):
+        self._logger.info(f"Increase reference for uris {uris}.")
+        for uri, _ in uris:
+            if uri in self._uri_reference:
+                self._uri_reference[uri] += 1
+            else:
+                self._uri_reference[uri] = 1
+
+    def _decrease_reference_for_uris(self, uris):
+        self._logger.info(f"Decrease reference for uris {uris}.")
+        unused_uris = list()
+        for uri, uri_type in uris:
+            if uri in self._uri_reference:
+                self._uri_reference[uri] -= 1
+                if self._uri_reference[uri] == 0:
+                    unused_uris.append((uri, uri_type))
+                    del self._uri_reference[uri]
+            else:
+                self._logger.error(f"URI {uri} does not exist.")
```

Comment:
Is it error? When does this usually happen?

Reply:
It happens when the caller(raylet) sends redundant PRC to delete runtime env. I also can modify it to warn.

Reply:
Can you tell me when the redundant RPCs can be sent to agent? what's the exact scenario?